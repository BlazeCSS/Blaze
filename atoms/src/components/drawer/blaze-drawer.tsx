import { Component, Prop, Method, State, Element } from '@stencil/core';

@Component({
  tag: 'blaze-drawer'
})
export class Drawer {

  @Element() elem: HTMLElement;
  @Prop() open: boolean = false;
  @Prop() dismissible: boolean = false;
  @Prop() position: string = 'bottom';
  @State() _isOpen: boolean = false;

  @Method()
  close() {
    this._isOpen = false;
  }

  @Method()
  show() {
    this._isOpen = true;
  }

  @Method()
  isOpen() {
    return this._isOpen;
  }

  componentWillLoad() {
    this._isOpen = this.open;
  }

  dismiss() {
    if (this.dismissible) this.close();
  }

  render() {
    const isOpenClass = this._isOpen ? 'o-drawer--visible' : '';
    const dismissibleOverlayClass = this.dismissible ? 'c-overlay--dismissible' : '';

    return (
      [
        this._isOpen && <div class={`c-overlay c-overlay--visible ${dismissibleOverlayClass}`} onClick={() => this.dismiss()}></div>,
        <div class={`o-drawer o-drawer--${this.position} ${isOpenClass}`}>
          <slot />
        </div>
      ]
    );
  }
}
