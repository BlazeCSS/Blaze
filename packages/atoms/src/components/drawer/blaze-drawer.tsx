import { Component, Prop, Method, State, Element } from '@stencil/core';

@Component({
  tag: 'blaze-drawer'
})
export class Drawer {

  @Element() elem: HTMLElement;
  @Prop() open: boolean = false;
  @Prop() dismissible: boolean = false;
  @Prop() position: string = 'bottom';
  @State() _isOpen: boolean = false;

  @Method()
  close() {
    this._isOpen = false;
  }

  @Method()
  show() {
    this._isOpen = true;
  }

  @Method()
  isOpen() {
    return this._isOpen;
  }

  componentWillLoad() {
    this._isOpen = this.open;
  }

  dismiss() {
    if (this.dismissible) this.close();
  }

  render() {
    const drawerIsOpenClass = this._isOpen ? 'o-drawer--visible' : '';

    return (
      [
        <blaze-overlay aria-hidden open={this._isOpen} onClick={() => this.dismiss()}></blaze-overlay>,
        <aside aria-expanded={this.isOpen} class={`o-drawer o-drawer--${this.position} ${drawerIsOpenClass}`}>
          <slot />
        </aside>
      ]
    );
  }
}
