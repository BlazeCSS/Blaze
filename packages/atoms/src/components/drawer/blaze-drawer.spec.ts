import { TestWindow } from '@stencil/core/dist/testing';
import { Drawer } from './blaze-drawer';

describe('Drawer', () => {
  it('should build', () => {
    expect(new Drawer()).toBeTruthy();
  });

  let element;

  const snapIt = (name, html) => {
    it(name, async () => {
      const window = new TestWindow();
      element = await window.load({
        components: [Drawer],
        html
      });
      window.flush();

      expect(element).toMatchSnapshot();
    });
  };

  snapIt(
    'renders correctly with no props',
    '<blaze-drawer>default</blaze-drawer>'
  );

  snapIt(
    'renders as open',
    '<blaze-drawer open>default</blaze-drawer>'
  );

  snapIt(
    'renders dismissible overlay',
    '<blaze-drawer open dismissible>default</blaze-drawer>'
  );

  it('show drawer method', async () => {
    const window = new TestWindow();
      element = await window.load({
      components: [Drawer],
      html: '<blaze-drawer>default</blaze-drawer>'
    });
    window.flush();
    expect(element).toMatchSnapshot();

    element.show();
    window.flush();
    expect(element.isOpen()).toEqual(true);
    expect(element).toMatchSnapshot();
  });

  it('close drawer method', async () => {
    const window = new TestWindow();
      element = await window.load({
      components: [Drawer],
      html: '<blaze-drawer open>default</blaze-drawer>'
    });
    window.flush();
    expect(element).toMatchSnapshot();

    element.close();
    window.flush();
    expect(element.isOpen()).toEqual(false);
    expect(element).toMatchSnapshot();
  });

  it('close by clicking overlay', async () => {
    const window = new TestWindow();
      element = await window.load({
      components: [Drawer],
      html: '<blaze-drawer open dismissible>default</blaze-drawer>'
    });
    window.flush();
    expect(element).toMatchSnapshot();

    element.querySelector('.c-overlay').click();
    window.flush();
    expect(element.isOpen()).toEqual(false);
    expect(element).toMatchSnapshot();
  });

  it('triggers onClose event', async (done) => {
    const window = new TestWindow();
    element = await window.load({
      components: [Drawer],
      html: `<blaze-drawer open dismissible>default</blaze-drawer>`
    });
    window.flush();

    element.addEventListener('onClose', () => {
      try {
        expect(element.isOpen()).toBe(false);
        done();
      } catch (err) {
        done.fail(err);
      }
    });

    expect(element.isOpen()).toBe(true);
    element.querySelector('.c-overlay').click();
  });
});
