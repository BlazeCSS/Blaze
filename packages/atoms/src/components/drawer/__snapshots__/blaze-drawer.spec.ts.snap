// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer close by clicking overlay 1`] = `
<blaze-drawer
  class="hydrated"
  dismissible=""
  open=""
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="true"
  />
  <aside
    class="o-drawer o-drawer--bottom o-drawer--visible"
  >
    
    default
  </aside>
</blaze-drawer>
`;

exports[`Drawer close by clicking overlay 2`] = `
<blaze-drawer
  class="hydrated"
  dismissible=""
  open=""
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="false"
  />
  <aside
    class="o-drawer o-drawer--bottom"
  >
    
    default
  </aside>
</blaze-drawer>
`;

exports[`Drawer close drawer method 1`] = `
<blaze-drawer
  class="hydrated"
  open=""
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="true"
  />
  <aside
    class="o-drawer o-drawer--bottom o-drawer--visible"
  >
    
    default
  </aside>
</blaze-drawer>
`;

exports[`Drawer close drawer method 2`] = `
<blaze-drawer
  class="hydrated"
  open=""
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="false"
  />
  <aside
    class="o-drawer o-drawer--bottom"
  >
    
    default
  </aside>
</blaze-drawer>
`;

exports[`Drawer renders as open 1`] = `
<blaze-drawer
  class="hydrated"
  open=""
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="true"
  />
  <aside
    class="o-drawer o-drawer--bottom o-drawer--visible"
  >
    
    default
  </aside>
</blaze-drawer>
`;

exports[`Drawer renders correctly with no props 1`] = `
<blaze-drawer
  class="hydrated"
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="false"
  />
  <aside
    class="o-drawer o-drawer--bottom"
  >
    
    default
  </aside>
</blaze-drawer>
`;

exports[`Drawer renders dismissible overlay 1`] = `
<blaze-drawer
  class="hydrated"
  dismissible=""
  open=""
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="true"
  />
  <aside
    class="o-drawer o-drawer--bottom o-drawer--visible"
  >
    
    default
  </aside>
</blaze-drawer>
`;

exports[`Drawer show drawer method 1`] = `
<blaze-drawer
  class="hydrated"
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="false"
  />
  <aside
    class="o-drawer o-drawer--bottom"
  >
    
    default
  </aside>
</blaze-drawer>
`;

exports[`Drawer show drawer method 2`] = `
<blaze-drawer
  class="hydrated"
>
  
  
  <blaze-overlay
    aria-hidden="true"
    open="true"
  />
  <aside
    class="o-drawer o-drawer--bottom o-drawer--visible"
  >
    
    default
  </aside>
</blaze-drawer>
`;
